pyro/infer/renyi_elbo.py

```diff
@@ -0,0 +1,179 @@
+from __future__ import absolute_import, division, print_function
+
+import math
+import warnings
+
+import pyro.poutine as poutine
+from pyro.distributions.util import is_identically_zero, log_sum_exp
+from pyro.infer.elbo import ELBO
+from pyro.infer.util import is_validation_enabled
+from pyro.poutine.util import prune_subsample_sites
+from pyro.util import check_model_guide_match, check_site_shape, torch_isnan
+
+
+class RenyiELBO(ELBO):
+    r"""
+    An implementation of Renyi's :math:`\alpha`-divergence variational inference
+    follows reference [1].
+
+    To have a lower bound, we require :math:`\alpha \ge 0`. However, according to
+    reference [1], depending on the dataset, :math:`\alpha < 0` might give better
+    results. In the special case :math:`\alpha = 0`, we have important weighted
+    lower bound derived in reference [2].
+
+    .. note:: Setting :math:`\alpha < 1` gives a better bound than the usual ELBO.
+        For :math:`\alpha = 1`, it is better to use
+        :class:`~pyro.infer.trace_elbo.Trace_ELBO` class because it helps reducing
+        variances of gradient estimations.
+
+    .. warning:: Mini-batch training is not supported yet.
+
+    :param float alpha: The order of :math:`\alpha`-divergence. Here
+        :math:`\alpha \neq 1`. Default is 0.
+    :param num_particles: The number of particles/samples used to form the ELBO
+        (gradient) estimators. Default is 2.
+    :param int max_iarange_nesting: Bound on max number of nested
+        :func:`pyro.iarange` contexts. Default is 2.
+    :param bool strict_enumeration_warning: Whether to warn about possible
+        misuse of enumeration, i.e. that
+        :class:`~pyro.infer.traceenum_elbo.TraceEnum_ELBO` is used iff there
+        are enumerated sample sites.
+
+    References:
+
+    [1] `Renyi Divergence Variational Inference`,
+        Yingzhen Li, Richard E. Turner
+
+    [2] `Importance Weighted Autoencoders`,
+        Yuri Burda, Roger Grosse, Ruslan Salakhutdinov
+    """
+
+    def __init__(self, alpha=0., num_particles=2, max_iarange_nesting=2,
+                 strict_enumeration_warning=True):
+        if alpha == 1:
+            raise ValueError("The order alpha should not be equal to 1. Please use Trace_ELBO class"
+                             "for the case alpha = 1.")
+        self.alpha = alpha
+        super(RenyiELBO, self).__init__(num_particles, max_iarange_nesting, vectorize_particles=True,
```

Comment:
I realized that it might be better to keep `vectorize_particles=False`, at least for now, because otherwise we will have to assume some value for `max_iarange_nesting` which is set to 2 here. This will cause dim clashes for models with more than 2 iaranges. We revert to True later, and set a higher default value for `max_iarange_nesting`, and throwing a sensible error if the model has more iaranges.